src/shared/tpm2-util.c

```diff
@@ -2248,8 +2248,36 @@ int tpm2_digest_many_digests(
  * the TPM specification Part 1 ("Architecture") section Authorization Values (subsection "Authorization Size
  * Convention") states "Trailing octets of zero are to be removed from any string before it is used as an
  * authValue". Since the TPM doesn't know if the auth value is a "string" or just a hash digest, any hash
- * digest that randomly happens to end in 0 must have the final 0 changed, or the TPM will remove it before
- * using the value in its HMAC calculations, resulting in failed HMAC checks. */
+ * digest that randomly happens to end in 0 must have the final 0(s) trimmed.
+ *
+ * This is required at 2 points. First, when setting the authValue during creation of new sealed objects, in
+ * tpm2_seal(). This only applies to newly created objects, of course.  Second, when using a previously
+ * created sealed object that has an authValue set, we use the sealed objects as the session bind key. This
+ * requires calling SetAuth so tpm2-tss can correctly calculate the HMAC to use for the encryption session.
+ *
+ * TPM implementations will perform the trimming for any authValue for existing sealed objects, so the
+ * tpm2-tss library must also perform the trimming before HMAC calculation, but it does not yet; this bug is
+ * open to add the trimming: https://github.com/tpm2-software/tpm2-tss/issues/2664
+ *
+ * Until our minimum tpm2-tss version contains a fix for that bug, we must perform the trimming
+ * ourselves. Note that since we are trimming, which is exactly what a TPM implementation would do, this will
+ * work for both existing objects with a authValue ending in 0(s) as well as new sealed objects we create,
+ * which we will trim the 0(s) from before sending to the TPM.
+ */
+static void tpm2_trim_auth_value(TPM2B_AUTH *auth) {
+        bool trimmed = false;
+
+        assert(auth);
+
+        while (auth->size > 0 && auth->buffer[auth.size - 1] == 0) {
+                trimmed = true;
+                auth->size--;
+        }
+
```

Comment:
maybe add an assert that auth->size is still > 0? Would be weird, but just to establish boundaries

Reply:
> maybe add an assert that auth->size is still > 0? Would be weird, but just to establish boundaries

Their is actually a possibility with poorly chosen auth values with a tpm, consider { 0, 0, 0, 0, 0 }, that auth value is going to size 0 after being trimmed. But it would be good to print a warning so folks know that maybe they should re-work their auth value? I was considering how to handle this in the tpm2-tss code, and was just considering a warning. With that said, maybe a hard failure is better so folks know they have a weak auth. But then you get into cases like { 1, 0, 0, 0, 0, 0, 0 } where they end up with a single digit for an auth value, and they may not have wanted that either... good grief TPM.

Reply:
no a 0-sized authValue is ok, in fact if the authValue buffer is all 0's that is exactly what should happen, the authValue size should be reduced to 0.

Reply:
I guess for our *specific* use case, which always is performing the trimming on a hash digest, there should be (effectively) no change at all of ever having an all-0 authValue. So we could add a assert or log warning if it is all 0's?

Reply:
> I guess for our _specific_ use case, which always is performing the trimming on a hash digest, there should be (effectively) no chance at all of ever having an all-0 authValue. So we could add a assert or log warning if it is all 0's?

Their should be effectively no change, but their is the incredibly slight chance that hash alg produces a digest of all zeros.
So a log warning would ensure that nothing completely breaks.